<a href="https://colab.research.google.com/github/dazzens/dazzens/blob/main/EDA_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

------------------
---------
-----

# **Пет проект EDA (исследовательский анализ данных).**




Автор: А.Г.Х.

Начало: 22.06.2026


Окончание: --.--.2026

----



##**Бизнес-контекст:**

- Интернет-магазин продает товары для дома и декора по всему миру. За последний год ассортимент вырос на 30%, но темпы роста выручки замедлились. Необходимо исследовать данные о продажах, чтобы понять, какие товары и страны приносят основную прибыль, кто самые ценные клиенты и как оптимизировать ценовую политику.

-----

# **Цель исследования**

- Выявить ключевые факторы, влияющие на динамику продаж интернет-магазина, для разработки рекомендаций по увеличению выручки и удержанию клиентов.


##**Вопросы:**


- 1. Как менялись продажи по месяцам? Динамика выручки, количества заказов и сезонность.


- 2. Какие товары самые популярные? Топ по количеству продаж, топ по выручке, топ по частоте заказов.

- 3. Какой средний чек и размер корзины? Среднее количество товаров в заказе, средняя сумма заказа

- 4. Какие страны приносят больше всего выручки? Топ по выручке, топ по количеству заказов, топ по среднему чеку

- 5. Кто самые прибыльные клиенты? Топ клиентов по сумме трат, по частоте покупок, анализ повторных покупок

- 6. Как цена влияет на продажи? Распределение цен, зависимость количества продаж от цены, оптимальные ценовые сегменты

- 7. Какие товары, страны и клиентские сегменты вносят наибольший вклад в выручку?


## **Содержание проекта**



1.  **Первичный осмотр и подготовка данных**
    *   Импорт библиотек и загрузка данных.
    *   Предварительный анализ структуры данных (типы, объем, пропуски).
    *   Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
    *   Проверка и обработка пропусков (`CustomerNo`).
    *   Проверка дубликатов и выбросов (`Price`, `Quantity`).

2.  **Предварительный анализ и создание новых признаков**
    *   Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
    *   Создание признаков: месяц, день недели, год из даты.
    *   Создание признака: общая сумма заказа (`Price` x `Quantity`).
    *   Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

3.  **Исследовательский анализ данных (EDA)**
    *   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
    *   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
    *   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
    *   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
    *   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
    *   Анализ ценовых сегментов: зависимость количества продаж от цены.

4.  **Ключевые выводы и рекомендации**
    *   Формирование основных инсайтов по каждому блоку анализа.
    *   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.


    

## **Описание данных:**

**Транзакционные данные (онлайн-магазин):**

| | | |
|------|----------|------------|
| `TransactionNo` | Уникальный идентификатор транзакции/заказа | `object` |
| `Date` | Дата совершения покупки | `object` |
| `ProductNo` | Уникальный идентификатор товара | `object` |
| `ProductName` | Название товара | `object` |
| `Price` | Цена за единицу товара | `float64` |
| `Quantity` | Количество единиц товара в заказе | `int64` |
| `CustomerNo` | Уникальный идентификатор клиента | `float64` |
| `Country` | Страна клиента | `object` |

---

**Характеристики данных:**

|  |  |
|------------|----------|
| **Объем данных** | 536 350 записей |
| **Количество клиентов** | 4 738 |
| **Количество уникальных товаров** | 3 768 |
| **Количество стран** | 38 |
| **Пропуски** | 55 записей без `CustomerNo` (0.01%) |



----
----

# **Этап 1 Первичный осмотр и подготовка данных**

- Импорт библиотек и загрузка данных.
- Предварительный анализ структуры данных (типы, объем, пропуски).
- Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
- Проверка и обработка пропусков (`CustomerNo`).
- Проверка дубликатов и выбросов (`Price`, `Quantity`).


---
## **Библиотеки**




In [1]:

# Необходимые библиотеки

try:
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    #from scipy.stats import ttest_ind
    #from matplotlib_venn import venn2
    import kagglehub
    import os
    print("Все библиотеки загружены")

except ImportError:
    !pip install numpy pandas matplotlib seaborn   kagglehub -q
    #scipy matplotlib-venn
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    #from scipy.stats import ttest_ind
    #from matplotlib_venn import venn2
    import kagglehub
    import os
    print("Библиотеки загружены через pip")



Все библиотеки загружены


## **Загрузка датасета:**

In [2]:
# Загруза датасета

path = kagglehub.dataset_download("gabrielramos87/an-online-shop-business")
file_path = os.path.join(path, [x for x in os.listdir(path) if x.endswith('.csv')][0])

df = pd.read_csv(file_path)

print(f" Датасет загружен: {df.shape}")

100%|██████████| 6.66M/6.66M [00:00<00:00, 109MB/s]

Extracting files...


 Датасет загружен: (536350, 8)


-----
----
----

## **Шаг 1. Общий осмотр данных**

In [3]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 536350 entries, 0 to 536349
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  536350 non-null  object 
 1   Date           536350 non-null  object 
 2   ProductNo      536350 non-null  object 
 3   ProductName    536350 non-null  object 
 4   Price          536350 non-null  float64
 5   Quantity       536350 non-null  int64  
 6   CustomerNo     536295 non-null  float64
 7   Country        536350 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.7+ MB


In [4]:
df.head(10)



,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
0,581482,12/9/2019,22485,Set Of 2 Wooden Market Crates,21.47,12,17490.0,United Kingdom
1,581475,12/9/2019,22596,Christmas Star Wish List Chalkboard,10.65,36,13069.0,United Kingdom
2,581475,12/9/2019,23235,Storage Tin Vintage Leaf,11.53,12,13069.0,United Kingdom
3,581475,12/9/2019,23272,Tree T-Light Holder Willie Winkie,10.65,12,13069.0,United Kingdom
4,581475,12/9/2019,23239,Set Of 4 Knick Knack Tins Poppies,11.94,6,13069.0,United Kingdom
5,581475,12/9/2019,21705,Bag 500g Swirly Marbles,10.65,24,13069.0,United Kingdom
6,581475,12/9/2019,22118,Joy Wooden Block Letters,11.53,18,13069.0,United Kingdom
7,581475,12/9/2019,22119,Peace Wooden Block Letters,12.25,12,13069.0,United Kingdom
8,581475,12/9/2019,22217,T-Light Holder Hanging Lace,10.65,12,13069.0,United Kingdom
9,581475,12/9/2019,22216,T-Light Holder White Lace,10.55,24,13069.0,United Kingdom


In [5]:
df.sample(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
50932,578028,11/22/2019,21935,Suki Shoulder Bag,11.94,1,17371.0,United Kingdom
425603,545872,3/7/2019,22557,Plasters In Tin Vintage Paisley,11.94,2,15752.0,United Kingdom
29648,579339,11/29/2019,22993,Set Of 4 Pantry Jelly Moulds,6.13,36,14866.0,United Kingdom
129340,572194,10/21/2019,85036C,Rose 1 Wick Morris Boxed Candle,11.53,1,17832.0,United Kingdom
402324,547964,3/28/2019,85131D,Beaded Crystal Heart Pink On Stick,11.10,2,14964.0,United Kingdom
121594,572918,10/26/2019,23300,Gardeners Kneeling Pad Cup Of Tea,11.94,1,14379.0,United Kingdom
256150,561795,7/29/2019,22343,Party Pizza Dish Red Retrospot,11.12,12,12716.0,France
485528,540604,1/10/2019,22715,Card Wedding Day,10.68,12,15326.0,United Kingdom
500666,539437,12/17/2018,21262,White Goose Feather Christmas Tree,18.06,1,13437.0,United Kingdom
198080,566949,9/15/2019,20718,Red Retrospot Shopper Bag,12.77,2,16949.0,United Kingdom


In [6]:
#предварительный осмотр
df.describe()


,Price,Quantity,CustomerNo
count,536350.000000,536350.000000,536295.000000
mean,12.662182,9.919347,15227.893178
std,8.490450,216.662300,1716.582932
min,5.130000,-80995.000000,12004.000000
25%,10.990000,1.000000,13807.000000
50%,11.940000,3.000000,15152.000000
75%,14.090000,10.000000,16729.000000
max,660.620000,80995.000000,18287.000000


### **Краткий вывод:**


---

## **Шаг 1.1 Обработка  типов данных**

In [14]:
#осмотр данных

df.dtypes

,0
TransactionNo,object
Date,object
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,float64
Country,object


In [15]:
# преобразование данных
df['Date'] = pd.to_datetime(df['Date'])
df['CustomerNo'] = df['CustomerNo'].astype('Int64')

#проверка
df.dtypes

,0
TransactionNo,object
Date,datetime64[ns]
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,Int64
Country,object


### **Итог преобразования:**
- Колонка `Date` преобразована в `datetime` для работы с временными рядами
- Колонка `CustomerNo` преобразована  в `Int64` так как содержит пропуски (`NaN`)

Данные оптимизированы для анализа

---
----

## **Шаг 1.2 Поиск и обработка пропусков**

In [16]:
# Пропуски: количество и доля
df.isnull().sum(), df.isnull().mean()

(TransactionNo     0
 Date              0
 ProductNo         0
 ProductName       0
 Price             0
 Quantity          0
 CustomerNo       55
 Country           0
 dtype: int64,
 TransactionNo    0.000000
 Date             0.000000
 ProductNo        0.000000
 ProductName      0.000000
 Price            0.000000
 Quantity         0.000000
 CustomerNo       0.000103
 Country          0.000000
 dtype: float64)

In [17]:
#проверка данных

df[df['CustomerNo'].isnull()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,<NA>,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,<NA>,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,<NA>,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,<NA>,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,<NA>,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,<NA>,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,<NA>,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,<NA>,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,<NA>,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,<NA>,United Kingdom


In [18]:
# Замена NaN на специальное значение
df['CustomerNo'] = df['CustomerNo'].fillna(-1)

In [19]:

# строки, где были пропуски (
display(df[df['CustomerNo'] == -1].head(10))

# 2. Проверка
print(f"\nВсего замен: {(df['CustomerNo'] == -1).sum()}")

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,-1,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,-1,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,-1,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,-1,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,-1,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,-1,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,-1,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,-1,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,-1,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,-1,United Kingdom



Всего замен: 55


### **Итог работы с пропусками:**

- Выбрана замена пустых значений для  сохранения данных



-----
-----

## **Шаг 1.3 Поиск и анализ  явных/неявных дубликатов**


### **Анализ явных дубликатов:**

In [20]:
#подсчет дубликатов

df.duplicated().sum()



np.int64(5200)

In [21]:
# Процент дубликатов
print(f"Процент дубликатов: {(df.duplicated().sum() / len(df) * 100):.2f}%")

# Первые 10 дубликатов
df[df.duplicated()].head(10)

Процент дубликатов: 0.97%


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
985,581497,2019-12-09,21481,Fawn Blue Hot Water Bottle,7.24,1,17497,United Kingdom
1365,581538,2019-12-09,23275,Set Of 3 Hanging Owls Ollie Beak,6.19,1,14446,United Kingdom
1401,581538,2019-12-09,22992,Revolver Wooden Ruler,6.19,1,14446,United Kingdom
1406,581538,2019-12-09,22694,Wicker Star,6.19,1,14446,United Kingdom
1409,581538,2019-12-09,23343,Jumbo Bag Vintage Christmas,6.19,1,14446,United Kingdom
3255,581237,2019-12-08,23498,Classic Bicycle Clips,6.19,1,15443,United Kingdom
3259,581237,2019-12-08,23497,Classic Chrome Bicycle Bell,6.19,1,15443,United Kingdom
3261,581237,2019-12-08,23509,Mini Playing Cards Fun Fair,6.19,2,15443,United Kingdom
3284,581237,2019-12-08,23378,Pack Of 12 50'S Christmas Tissues,6.19,1,15443,United Kingdom
3593,581253,2019-12-08,23427,Stool Home Sweet Home,6.19,1,16891,United Kingdom


**Процент малый повторов, можно смело удалять**

In [22]:
#удаление дубликатов и проверка


df = df.drop_duplicates()

df.duplicated().sum()

np.int64(0)

- Дубликатов явных не осталось в датасете

----

### **Анализ неявных дубликатов:**

In [23]:
# Показать реальные дубликаты (первые 10)
df[df.duplicated(keep=False)].sort_values('TransactionNo').head(20)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country


In [24]:
# ============================================
# АНАЛИЗ НЕЯВНЫХ ДУБЛИКАТОВ
# ============================================

# 1. Находим заказы с повторяющимися товарами
duplicate_check = df.groupby(['TransactionNo', 'ProductNo']).size().reset_index(name='count')
implicit_duplicates = duplicate_check[duplicate_check['count'] > 1]

print(f"Найдено неявных дубликатов: {len(implicit_duplicates)}")
display(implicit_duplicates.head(10))

# 2. Показываем пример таких строк
print("\nПример строк с неявными дубликатами:")
example = df[df['TransactionNo'].isin(implicit_duplicates['TransactionNo'].head(3))]
display(example[['TransactionNo', 'ProductNo', 'ProductName', 'Quantity']].sort_values(['TransactionNo', 'ProductNo']))

Найдено неявных дубликатов: 5553


,TransactionNo,ProductNo,count
133,536381,71270,2
515,536409,85116,2
518,536409,90199C,3
529,536412,21448,2
535,536412,21738,3
538,536412,22077,3
541,536412,22144,2
545,536412,22243,3
547,536412,22273,2
562,536412,22749,2



Пример строк с неявными дубликатами:


,TransactionNo,ProductNo,ProductName,Quantity
533393,536381,15056N,Edwardian Parasol Natural,2
533379,536381,21166,Cook With Wine Metal Sign,1
533378,536381,21169,You're Confusing Me Metal Sign,3
533380,536381,21175,Gin And Tonic Diet Metal Sign,2
533400,536381,21523,Doormat Fancy Font Home Sweet Home,10
...,...,...,...,...
533731,536409,90129F,Red Glass Tassle Bag Charm,1
533718,536409,90199C,5 Strand Glass Necklace Crystal,3
533733,536409,90199C,5 Strand Glass Necklace Crystal,1
533759,536409,90199C,5 Strand Glass Necklace Crystal,2


In [25]:
# ============================================
# ТОЛЬКО НЕЯВНЫЕ ДУБЛИКАТЫ (повторы)
# ============================================

# 1. Находим товары, которые повторяются в заказе
duplicate_check = df.groupby(['TransactionNo', 'ProductNo']).size().reset_index(name='count')
duplicate_check = duplicate_check[duplicate_check['count'] > 1]

# 2. Показываем только те строки, которые являются повторами
print(f"Найдено неявных дубликатов: {len(duplicate_check)}")
display(duplicate_check.head(10))

# 3. Показываем пример повторов с датой
print("\nПример неявных дубликатов (повторяющиеся товары):")
duplicate_transactions = duplicate_check['TransactionNo'].head(3)
for trans in duplicate_transactions:
    print(f"\nЗаказ: {trans}")
    display(df[df['TransactionNo'] == trans][['TransactionNo', 'Date', 'ProductNo', 'ProductName', 'Quantity']])

Найдено неявных дубликатов: 5553


,TransactionNo,ProductNo,count
133,536381,71270,2
515,536409,85116,2
518,536409,90199C,3
529,536412,21448,2
535,536412,21738,3
538,536412,22077,3
541,536412,22144,2
545,536412,22243,3
547,536412,22273,2
562,536412,22749,2



Пример неявных дубликатов (повторяющиеся товары):

Заказ: 536381


,TransactionNo,Date,ProductNo,ProductName,Quantity
533367,536381,2018-12-01,22139,Retrospot Tea Set Ceramic 11 Pc,23
533368,536381,2018-12-01,84854,Girly Pink Tool Set,5
533369,536381,2018-12-01,22411,Jumbo Shopper Vintage Red Paisley,10
533370,536381,2018-12-01,82567,Airline Lounge Metal Sign,2
533371,536381,2018-12-01,21672,White Spot Red Ceramic Drawer Knob,6
533372,536381,2018-12-01,22774,Red Drawer Knob Acrylic Edwardian,24
533373,536381,2018-12-01,22771,Clear Drawer Knob Acrylic Edwardian,24
533374,536381,2018-12-01,71270,Photo Clip Line,1
533375,536381,2018-12-01,22262,Felt Egg Cosy Chicken,1
533376,536381,2018-12-01,22637,Piggy Bank Retrospot,1



Заказ: 536409


,TransactionNo,Date,ProductNo,ProductName,Quantity
533718,536409,2018-12-01,90199C,5 Strand Glass Necklace Crystal,3
533719,536409,2018-12-01,21479,White Skull Hot Water Bottle,1
533720,536409,2018-12-01,22111,Scottie Dog Hot Water Bottle,1
533721,536409,2018-12-01,22785,Squarecushion Cover Pink Union Jack,1
533722,536409,2018-12-01,22975,Spaceboy Childrens Egg Cup,1
533723,536409,2018-12-01,22972,Children's Spaceboy Mug,1
533724,536409,2018-12-01,22866,Hand Warmer Scotty Dog Design,1
533725,536409,2018-12-01,22568,Feltcraft Cushion Owl,1
533726,536409,2018-12-01,85116,Black Candelabra T-Light Holder,1
533727,536409,2018-12-01,22664,Toy Tidy Dolly Girl Design,1



Заказ: 536409


,TransactionNo,Date,ProductNo,ProductName,Quantity
533718,536409,2018-12-01,90199C,5 Strand Glass Necklace Crystal,3
533719,536409,2018-12-01,21479,White Skull Hot Water Bottle,1
533720,536409,2018-12-01,22111,Scottie Dog Hot Water Bottle,1
533721,536409,2018-12-01,22785,Squarecushion Cover Pink Union Jack,1
533722,536409,2018-12-01,22975,Spaceboy Childrens Egg Cup,1
533723,536409,2018-12-01,22972,Children's Spaceboy Mug,1
533724,536409,2018-12-01,22866,Hand Warmer Scotty Dog Design,1
533725,536409,2018-12-01,22568,Feltcraft Cushion Owl,1
533726,536409,2018-12-01,85116,Black Candelabra T-Light Holder,1
533727,536409,2018-12-01,22664,Toy Tidy Dolly Girl Design,1


### **кратный вывод**




----
----

## **Шаг 2 Детальный осмотр данных:**

In [7]:
display(df[df['Quantity'] < 0].head(15).sort_values(by='Quantity'))

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
1616,C581484,12/9/2019,23843,Paper Craft Little Birdie,6.19,-80995,16446.0,United Kingdom
6480,C581228,12/8/2019,22467,Gumball Coat Rack,6.19,-36,16019.0,United Kingdom
6484,C581229,12/8/2019,23158,Set Of 5 Lucky Cat Magnets,6.19,-36,12558.0,USA
6481,C581229,12/8/2019,84828,Jungle Popsicles Ice Lolly Moulds,6.19,-24,12558.0,USA
6479,C581228,12/8/2019,22781,Gumball Magazine Rack,6.19,-24,16019.0,United Kingdom
6483,C581229,12/8/2019,23392,Spaceboy Rocket Lolly Makers,6.19,-24,12558.0,USA
6482,C581229,12/8/2019,79191C,Retro Plastic Elephant Tray,6.19,-24,12558.0,USA
1617,C581490,12/9/2019,22178,Victorian Glass Hanging T-Light,6.19,-12,14397.0,United Kingdom
6477,C581228,12/8/2019,23210,White Rocking Horse Hand Painted,6.19,-12,16019.0,United Kingdom
1618,C581490,12/9/2019,23144,Zinc T-Light Holder Stars Small,6.04,-11,14397.0,United Kingdom


In [8]:
display(df[df['Price'] > 27].head(15).sort_values(by='Price', ascending=False))

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
49340,578042,11/22/2019,22826,Love Seat Antique White Metal,97.37,6,17500.0,United Kingdom
12297,580730,12/5/2019,22847,Bread Bin Diner Style Ivory,44.37,1,15730.0,United Kingdom
12296,580730,12/5/2019,22846,Bread Bin Diner Style Red,44.37,1,15730.0,United Kingdom
49342,578065,11/22/2019,22423,Regency Cakestand 3 Tier,35.83,2,15065.0,United Kingdom
49343,578065,11/22/2019,22622,Box Of Vintage Alphabet Blocks,35.83,1,15065.0,United Kingdom
49341,577844,11/22/2019,22423,Regency Cakestand 3 Tier,35.83,13,16844.0,United Kingdom
49344,578067,11/22/2019,22423,Regency Cakestand 3 Tier,35.83,3,12067.0,United Kingdom
49345,578067,11/22/2019,23553,Landmark Frame Camden Town,35.83,1,12067.0,United Kingdom
12298,580730,12/5/2019,23173,Regency Teapot Roses,30.71,2,15730.0,United Kingdom
12300,580730,12/5/2019,22626,Black Kitchen Scales,27.30,1,15730.0,United Kingdom


In [9]:
# промежуточный анализ


sum_quantity_min = (df['Quantity'] < 0).sum()
count_quantity_min = (df['Quantity'] < 0).count()

print(sum_quantity_min)
print(count_quantity_min)

sum_negative = df[df['Quantity'] < 0]['Quantity'].sum()
print(f"Сумма отрицательных Quantity: {sum_negative}")


#######
######
#подсчет отрицательных колонок


negative_count = (df['Quantity'] < 0).sum()
print(f"Отрицательных Quantity: {negative_count:}")

# Доля отрицательных Quantity в процентах
negative_pct = (df['Quantity'] < 0).mean() * 100
print(f"Доля отрицательных Quantity: {negative_pct:.2f}%")

8585
536350
Сумма отрицательных Quantity: -271256
Отрицательных Quantity: 8585
Доля отрицательных Quantity: 1.60%


# **// Понять природу отрицательного числа количества**

// Cделать предположение почему оно появилосб и дополнить краткий вывод на предосмотре

### **Осмотр уникальных позиций**

In [10]:
# Быстрый взгляд на все категориальные поля
df[['Country',
    'CustomerNo',
    'ProductName',
    'TransactionNo']].nunique().sort_values(ascending=False)


,0
TransactionNo,23204
CustomerNo,4738
ProductName,3768
Country,38


### **Анализ хвостов и подсчет хваостов**

- нужно подсчитать что выше 99 квартиля и какая сумма формируется от общего число и понять выброс это или нет

In [11]:
#Анализ хвоста цены

df['Price'].quantile([0.25,
                      0.5,
                      0.75,
                      0.9,
                      0.95,
                      0.99])




,Price
0.25,10.99
0.50,11.94
0.75,14.09
0.90,17.37
0.95,19.42
0.99,27.30


In [12]:

df['Quantity'].quantile([0.25,
                      0.5,
                      0.75,
                      0.9,
                      0.95,
                      0.99])

,Quantity
0.25,1.0
0.50,3.0
0.75,10.0
0.90,24.0
0.95,29.0
0.99,100.0


In [13]:
#Анализ 99+ процентиля
upper_99 = df['Price'].quantile(0.99);
high_outliers = df[df['Price'] > upper_99]


display(high_outliers.sort_values(by='Price', ascending=False))

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
432039,545158,2/28/2019,22656,Vintage Blue Kitchen Cabinet,660.62,1,15874.0,United Kingdom
531163,536835,12/2/2018,22655,Vintage Red Kitchen Cabinet,660.62,1,13145.0,United Kingdom
499322,C539438,12/17/2018,22655,Vintage Red Kitchen Cabinet,660.62,-1,16607.0,United Kingdom
499323,539440,12/17/2018,22655,Vintage Red Kitchen Cabinet,660.62,2,16607.0,United Kingdom
484702,540647,1/10/2019,22655,Vintage Red Kitchen Cabinet,660.62,1,17406.0,United Kingdom
...,...,...,...,...,...,...,...,...
130213,572049,10/20/2019,23486,Antique Heart Shelf Unit,27.32,2,17243.0,United Kingdom
130214,572093,10/20/2019,23486,Antique Heart Shelf Unit,27.32,1,15121.0,United Kingdom
130215,572098,10/20/2019,23486,Antique Heart Shelf Unit,27.32,1,15117.0,United Kingdom
130216,572103,10/20/2019,23486,Antique Heart Shelf Unit,27.32,1,17571.0,United Kingdom


----
----
### // **На доработку Краткий вывод**

В результате предварительно осмотра установлено, что:

- Объём выборки: 536 350 записей, 8 признаков.
- Пропуски: 55 `CustomerNo`
- Размерность данных: 23 204 транзакции, 4 738 клиентов, 3 768 товаров, 38 стран.
- Распределение цен: медиана — 11.94 руб., 90-й перцентиль — 17.37 руб., 99-й — 27.30 руб.


Набдюдаются следующие аномалии :
- отрицательное числа в  `Quantity`  
- крупное число позиций в `Quantity`
- в колонке `price`  завешенная цена **660** относительно других цен

Неоходимо выявить причину возникновения

-----
-----

#// **Вот тут можно начать анализировть отрицательное количество и дальше думать что с ним делать**

----
----
----

#**Этап 2. Предварительный анализ и создание новых признаков**

- Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
- Создание признаков: месяц, день недели, год из даты.
- Создание признака: общая сумма заказа (`Price` и `Quantity`).
- Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

# // **Создать колонку где будут лишь положительное колличество цен**

# **Этап 3. Исследовательский анализ данных (EDA)**

*   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
*   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
*   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
*   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
*   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
*   Анализ ценовых сегментов: зависимость количества продаж от цены.

In [26]:
#топ-10 дорогих товаров

top_price = df.groupby('ProductName')['Price'].max().sort_values(ascending=False).head(10)
print(top_price)

ProductName
Vintage Red Kitchen Cabinet          660.62
Vintage Blue Kitchen Cabinet         660.62
Love Seat Antique White Metal        210.12
Rustic Seventeen Drawer Sideboard    179.37
Regency Mirror With Shutters         179.37
Chest Natural Wood 20 Drawers        138.38
Hall Cabinet With 3 Drawers          112.75
Sweetheart Cream Steel Table Rect    112.71
Cinderella Chandelier                112.71
Ivory Embroidered Quilt               94.81
Name: Price, dtype: float64


# **Этап 4. Ключевые выводы и рекомендации**
*   Формирование основных инсайтов по каждому блоку анализа.
*   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.